packages/hierarchy-builder/src/hierarchy-builder/TreeNodesProvider.ts

```diff
@@ -0,0 +1,515 @@
+            // will process each observable that is present in the pipeline recursively.
+            observeOn(queueScheduler),
+            // Delay the connection until another event loop task
+            subscribeOn(asapScheduler),
+            // Ignore errors in this pipeline without suppressing them for other subscribers
+            onErrorResumeNext,
+          );
+        }, QUERY_CONCURRENCY),
+      )
+      // Start consuming scheduled observables
+      .subscribe();
+  }
+  /**
+   * Schedules `source` for subscription in the current scheduler.
+   *
+   * The actual scheduling is performed when the returned observable is subscribed to. To cancel, remove all subscribers
+   * from the returned observable.
+   *
+   * @param source Input observable for which to schedule a subscription.
+   * @returns Hot observable which starts emitting `source` values after subscription.
+   */
+  public scheduleSubscription(source: Observable<T>): Observable<T> {
+    return defer(() => {
+      let unsubscribed = false;
+      const connectableObservable = connectable(iif(() => unsubscribed, EMPTY, source));
+      this._scheduler.next(connectableObservable);
+      return connectableObservable.pipe(
+        finalize(() => {
+          unsubscribed = true;
+        }),
+      );
+    });
+  }
+}
+
+export interface TreeNodesProviderProps {
+  schemas: SchemaContext;
+  queryBuilder: ITreeQueryBuilder;
+  queryExecutor: IQueryExecutor;
+}
+
+export class TreeNodesProvider {
+  private _schemas: SchemaContext;
+  private _queryBuilder: ITreeQueryBuilder;
+  private _queryExecutor: IQueryExecutor;
+  private _queryReader: TreeQueryResultsReader;
+  private _scheduler: QueryScheduler<InProgressTreeNode[]>;
+  private _directNodesCache: Map<string, Observable<InProgressTreeNode>>;
+
+  public constructor(props: TreeNodesProviderProps) {
+    this._schemas = props.schemas;
+    this._queryBuilder = props.queryBuilder;
+    this._queryExecutor = props.queryExecutor;
+    this._queryReader = new TreeQueryResultsReader();
+    this._scheduler = new QueryScheduler();
+    this._directNodesCache = new Map();
+  }
+
+  private loadDirectNodes(parentNode: TreeNode | undefined): Observable<InProgressTreeNode> {
+    const enableLogging = false;
+    return from(this._queryBuilder.createQueries(parentNode)).pipe(
+      mergeMap((queries) => from(queries)),
+      (queryObservable) =>
+        this._scheduler.scheduleSubscription(
+          queryObservable.pipe(
+            tap(() => enableLogging && console.log(`[loadDirectNodes] Do query for ${parentNode ? JSON.stringify(parentNode) : "<root>"}`)),
+            mergeMap((query) => from(this._queryReader.read(this._queryExecutor, { ...query, ecsql: applyLimit(query.ecsql, query.ctes) }))),
+          ),
+        ),
+      mergeAll(),
+      map(convertECInstanceIdSuffixToBase36),
+      shareReplay({ refCount: false }),
```

Comment:
Looks like `refCount: false` is default value https://rxjs.dev/api/index/function/shareReplay#reference-counting. Maybe this can be removed

Reply:
Cleaned up